[FLOC-4203] Out of process test runner

lukemarsden opened the pull request:
Fixes https://clusterhq.atlassian.net/browse/FLOC-4203

This currently runs the out of process test runner against the Python `dvol` implementation, but we'll soon be able to swap that out for the beginnings of a golang one.

jml commented on dvol.go:
```diff
@@ -0,0 +1,16 @@
+    "fmt"
+
+//    "github.com/clusterhq/dvol/dockercontainers"
+//    "github.com/clusterhq/dvol/plugin"
```
I recommend deleting all of the Go stuff from this change and just focusing on the out-of-process stuff. The Go bits can be in the next patch.

jml commented on dvol_python/dvol.py:
```diff
@@ -88,11 +101,13 @@ def write(self, volume, branch, commitData):
+        self.lockFactory = DockerLock
+        if disable_docker_integration:
+            # Do not attempt to connect to Docker if we've been asked not to.
+            self.lockFactory = NullLock
```
Just pass in the lock factory. Booleans don't mean anything by themselves, whereas the lock factory does.

jml commented on dvol_python/test_dvol.py:
```diff
@@ -126,6 +161,7 @@ def test_list_multi_volumes(self, volumes):
             sorted([line.split() for line in rest]),
         )
 
+    @settings(max_examples=5)
```
Rather than doing this here, which implies that the `max_examples` is somehow intrinsic to the test, instead put something like this in the `tests/__init__.py` file:

```python
import os
from hypothesis import settings

settings.register_profile("ci", settings(max_examples=1000))
settings.register_profile("dev", settings(max_examples=5))
settings.register_profile("debug", settings(max_examples=10, verbosity=Verbosity.verbose))

settings.load_profile(os.environ.get(u'HYPOTHESIS_PROFILE', 'default'))
```

And then on your laptop (perhaps in an `.envrc`?) set `HYPOTHESIS_PROFILE=dev`.

This might seem like gold-plating, but if we're sticking with Python tests, it'll be worth it.

jml commented:
Thanks Luke!

Since (I think) this is the first time I've reviewed your code, I'd like to have another look before this lands.

lukemarsden commented:
Thanks for the review, @jml! I've addressed the feedback, please take another look.

One thing I wasn't sure about is whether having 1000 examples per hypothesis test in CI is a good idea. Having fast CI is about as important as having fast dev cycles, IMO. So I didn't set the environment variable in the `.travis.yml`, and I set the default to `dev` rather than `default`. Let me know what you think.

Cheers,
Luke

jml commented:
Thanks! 

I still think we should use an explicit, separate env var for CI examples. If 1,000 examples takes too long (how long does it take?) then we can lower it.

jml commented:
Please address optional comment & merge.

jml commented:
Please address optional comment & merge.

lukemarsden commented:
Thanks! Merging.
